python/lsst/ts/observatory/control/base_group.py

```diff
@@ -0,0 +1,303 @@
+        return salobj.State(ss.summaryState)
+
+    async def check_component_state(self, comp, desired_state=salobj.State.ENABLED):
+        """Given a component name wait for an event that specify that the
+        state changes. If the event is not ENABLED, raise RuntimeError.
+
+        This is to be used in conjunction with the `get_state_changed_callback`
+        method.
+
+        Parameters
+        ----------
+        comp : `str`
+            Name of the component to follow. Must be one of:
+            atmcs, atptg, ataos, atpneumatics, athexapod, atdome,
+            atdometrajectory
+        desired_state : `salobj.State`
+            Desired state of the CSC.
+
+        Raises
+        ------
+        RuntimeError
+            If state is not `desired_state`.
+        """
+        first_pass = True
+        while True:
+
+            if first_pass:
+                _state = await self._remotes[comp].evt_summaryState.aget()
+                self._remotes[comp].evt_summaryState.flush()
+                first_pass = False
+            else:
+                try:
+                    _state = await self._remotes[comp].evt_summaryState.next(
+                        flush=False
+                    )
+                except IndexError:
+                    _state = await self._remotes[comp].evt_summaryState.aget()
+
+            state = salobj.State(_state.summaryState)
+
+            if state != desired_state:
+                self.log.warning(f"{comp} not in {desired_state!r}: {state!r}")
+                raise RuntimeError(
+                    f"{comp} state is {state!r}, " f"expected {desired_state!r}"
+                )
+            else:
+                self.log.debug(f"{comp}: {state!r}")
+
+    async def check_comp_heartbeat(self, comp):
+        """
+
+        Parameters
+        ----------
+        comp
+
+        Returns
+        -------
+
+        """
+
+        while True:
+            await self._remotes[comp].evt_heartbeat.next(
+                flush=True, timeout=self.fast_timeout
+            )
+
+    async def standby(self):
+        """ Put all CSCs in standby.
+        """
+
+        set_ss_tasks = []
+
+        for comp in self.components:
+            if getattr(self.check, comp):
+                set_ss_tasks.append(
+                    salobj.set_summary_state(
+                        self._remotes[comp],
+                        salobj.State.STANDBY,
+                        timeout=self.long_long_timeout,
+                    )
+                )
+            else:
+                set_ss_tasks.append(self.get_state(comp))
+
+        ret_val = await asyncio.gather(*set_ss_tasks, return_exceptions=True)
+
+        error_flag = False
+        error_msg = ""
+
+        for i in range(len(self.components)):
+            if isinstance(ret_val[i], Exception):
+                error_flag = True
+                error_msg += f"Unable to put {self.components[i]} in STANDBY\n"
```

Comment:
error_msg will be needlessly repetitive. Also you are mis-calling log.exception.

See my comment below (in the `enable` method) for details and suggested alternate code that accumulates the failed component names in a list instead of having error_flag and error_msg.